### PR TITLE
18dixie preferred shares

### PR DESCRIPTION
**Pull request by ventusignis**
18Dixie minors are exchanged for preferred shares which do not count towards the 60%-of-a-corp limit but are otherwise 10% shares

* Uses the share_flags view introduced in 21moon for labelling preferred shares; refactored 21moon so that share_flags is a game level method, not step level

* Added `group_by(&:percent)` to the share sizes parameter for the buy button, without this change the market share options would say "10%" share which isn't needed

image

Minor closure is implemented as a merger round. Currently in the case for minors where there is only one choice of share to exchange for, the player must click exchange

More pictures
image
image
image

### CODE REVIEW

**Comment by tobymao** [comment]
@roseundy can you review this?

File: assets/app/view/game/corporation.rb [hunk 1]
```diff
@@ -413,8 +413,7 @@ def render_shares
         if player_rows.any? || @corporation.num_market_shares.positive?
           at_limit = @game.share_pool.bank_at_limit?(@corporation)
           double_certs = @game.share_pool.shares_of(@corporation).count(&:double_cert)
-          step = @game.round.active_step
-          other_flags = step.respond_to?(:share_flags) && step&.share_flags(@game.share_pool.shares_of(@corporation))
+          other_flags = @game.respond_to?(:share_flags) && @game.share_flags(@game.share_pool.shares_of(@corporation))
```
Comment (roseundy) [inline assets/app/view/game/corporation.rb:416]: If you are going to add a @game method, you don't need to use respond_to?, instead add it to the Game::base class.

File: lib/engine/game/g_21_moon/game.rb [hunk 2]
```diff
@@ -1034,6 +1034,11 @@ def map_legend
             ],
           ]
         end
+
+        def share_flags(shares)
+          step = @round.active_step
+          step.share_flags(shares) if step.respond_to?(:share_flags)
+        end
```
Comment (roseundy) [inline lib/engine/game/g_21_moon/game.rb:1041]: As I mentioned above, just move the share_flags method from the step to here.

File: lib/engine/share.rb [hunk 3]
```diff
@@ -63,6 +66,12 @@ def inspect
       "<Share: #{@corporation.id} #{@percent}%>"
     end
 
+    def preferred_percent
+      return 0 if @preferred
+
+      @percent
+    end
```
Comment (roseundy) [inline lib/engine/share.rb:73]: The naming seems a little strange. This says that preferred_percent is zero if it's preferred?

Shouldn't this be "common_percent" ?

File: lib/engine/share_bundle.rb [hunk 4]
```diff
@@ -65,5 +65,10 @@ def buyable
     def ==(other)
       [shares, percent, share_price] == [other.shares, other.percent, other.share_price]
     end
+
+    # Percent of shares not counting preferred shares
+    def preferred_percent
+      @preferred_percent ||= @shares.reject(&:preferred).sum(&:percent)
+    end
```
Comment (roseundy) [inline lib/engine/share_bundle.rb:72]: common_percent

File: lib/engine/share_pool.rb [hunk 5]
```diff
@@ -154,7 +154,7 @@ def fit_in_bank?(bundle)
     end
 
     def bank_at_limit?(corporation)
-      percent_of(corporation) >= @game.class::MARKET_SHARE_LIMIT
+      preferred_percent_of(corporation) >= @game.class::MARKET_SHARE_LIMIT
```
Comment (roseundy) [inline lib/engine/share_pool.rb:157]: common_percent_of

File: lib/engine/step/share_buying.rb [hunk 6]
```diff
@@ -39,7 +39,7 @@ def can_gain?(entity, bundle, exchange: false)
 
         corporation = bundle.corporation
 
-        corporation.holding_ok?(entity, bundle.percent) &&
+        corporation.holding_ok?(entity, bundle.preferred_percent) &&
           (!corporation.counts_for_limit || exchange || @game.num_certs(entity) < @game.cert_limit)
```
Comment (roseundy) [inline lib/engine/step/share_buying.rb:43]: common_percent

**Comment by roseundy** [comment]
I'm a little nit-picky about the naming of methods in the shared code.

File: lib/engine/share.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/share.rb:73]: I see what you mean by this; that's a bit clearer than what I was thinking (preferred_percent -> percent accounting for preferred status)

File: assets/app/view/game/corporation.rb (hunk 1 above)
Comment (ventusignis) [reply assets/app/view/game/corporation.rb]: Done

File: lib/engine/share.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/share.rb:75]: Done

File: lib/engine/share_bundle.rb (hunk 4 above)
Comment (ventusignis) [reply lib/engine/share_bundle.rb:72]: Done

File: lib/engine/share_holder.rb [hunk 7]
```diff
@@ -32,6 +32,13 @@ def percent_of(corporation)
       shares_by_corporation[corporation].sum(&:percent)
     end
 
+    # Same as percent_of, except preferred shares don't count
+    def preferred_percent_of(corporation)
```
[the remark this answers is not in the document]
Comment (ventusignis) [reply lib/engine/share_holder.rb]: Done

File: lib/engine/share_pool.rb (hunk 5 above)
Comment (ventusignis) [reply lib/engine/share_pool.rb]: Done

File: lib/engine/step/share_buying.rb (hunk 6 above)
Comment (ventusignis) [reply lib/engine/step/share_buying.rb:43]: Done

**Comment by ventusignis** [comment]
I'm all for nit-picking naming of methods in common code; if it's going to be in the common code it's intended to be used by other games & developers and it should be easy to understand what's happening

File: assets/app/view/game/buy_sell_shares.rb [hunk 8]
```diff
@@ -86,7 +86,7 @@ def render_market_shares
           h(Button::BuyShare,
             share: share,
             entity: @current_entity,
-            percentages_available: @pool_shares.size)
+            percentages_available: @pool_shares.group_by(&:percent).size)
```
Comment (roseundy) [inline assets/app/view/game/buy_sell_shares.rb:89]: Why is this needed? Haven't these already been grouped by percent?

**Review by roseundy — APPROVED** [review]
